<!DOCTYPE html>
<html lang="en">
    <head>
        <link href="css/flat-ui.css" rel="stylesheet">
        <script>
          (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
          (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
          m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
          })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

          ga('create', 'UA-47481907-2', 'iobio.io');
          ga('send', 'pageview');

        </script>

        <script src="js/jquery.min.js"></script>
        <script src="js/class.js"></script>
        <script src="js/rdp.js"></script>
        <script src="js/bootstrap/js/bootstrap.min.js"></script>
        <script src="js/d3.min.js"></script>
        <script src="js/queue.min.js"></script>
        <script src="js/donut.d3.js"></script>
        <script src="js/histogram.d3.js"></script>
        <script src="js/histogramViewFinder.d3.js"></script>
        <script src="js/movingLine.d3.js"></script>
        <script src="js/nprogress.js"></script>
        <script src="js/bam.d3.js"></script>
        <script src="js/bam.iobio.js/bam.iobio.js"></script>
        <script src="js/bam.iobio.js/bam.js"></script>
        <script src="js/bam.iobio.js/bin.js"></script>
        <script src="js/bam.iobio.js/inflate.js"></script>

        <script src="js/iobio.viz.js"></script>
        <script src="js/iobio.js"></script>
        <link rel="stylesheet" href="js/bootstrap/css/bootstrap.min.css">
        <link href='http://fonts.googleapis.com/css?family=Quicksand:300,400' rel='stylesheet' type='text/css'>
        <link href='http://fonts.googleapis.com/css?family=Overlock+SC' rel='stylesheet' type='text/css'>
        <link href="https://fonts.googleapis.com/css?family=Open+Sans" rel="stylesheet">
        <link href='css/nprogress.css' rel='stylesheet' type='text/css'>
        <link href='css/site.css' rel='stylesheet' type='text/css'>
        <link href='css/iobio.viz.min.css' rel='stylesheet' type='text/css'>
        <!-- <link href='http://fonts.googleapis.com/css?family=Quicksand:300,400' rel='stylesheet' type='text/css'> -->

        <script>
            $(document).ready( function(){
                // var is_chrome = navigator.userAgent.toLowerCase().indexOf('chrome') > -1;
                // if (!is_chrome) alert("Warning! This site requires Google Chrome to function properly");
                // check for window.requestAnimation frame and add it if it's not here.
                (function () {
                  var lastTime = 0;
                  var vendors = ['ms', 'moz', 'webkit', 'o'];
                  for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
                    window.requestAnimationFrame = window[vendors[x] + 'RequestAnimationFrame'];
                    window.cancelAnimationFrame = window[vendors[x] + 'CancelAnimationFrame'] || window[vendors[x] + 'CancelRequestAnimationFrame'];
                  }
                  if(!window.requestAnimationFrame)
                    window.requestAnimationFrame = function (callback, element) {
                      var currTime = new Date().getTime();
                      var timeToCall = Math.max(0, 16 - (currTime - lastTime));
                      var id = window.setTimeout(function () {
                        callback(currTime + timeToCall);
                      },
                      timeToCall);
                      lastTime = currTime + timeToCall;
                      return id;
                  };
                  if(!window.cancelAnimationFrame)
                    window.cancelAnimationFrame = function (id) {
                      clearTimeout(id);
                  };
                }());

                // Help modals
                $("#genome_coverage_help").click(function() {
                  $("#genomeCoverageModal").modal({
                    show: true
                  });
                });

                $("#sampled_help").click(function() {
                  $("#sampledModal").modal({
                    show: true
                  });
                });

                $("#mapped_reads_help").click(function() {
                  $("#mappedReadsModal").modal({
                    show: true
                  });
                });

                $("#strands_help").click(function() {
                  $("#strandsModal").modal({
                    show: true
                  });
                });

                $("#proper_pairs_help").click(function() {
                  $("#properPairsModal").modal({
                    show: true
                  });
                });

                $("#singletons_help").click(function() {
                  $("#singletonsModal").modal({
                    show: true
                  });
                });

                $("#both_mapped_help").click(function() {
                  $("#bothMappedModal").modal({
                    show: true
                  });
                });

                $("#duplicates_help").click(function() {
                  $("#duplicatesModal").modal({
                    show: true
                  });
                });

                $("#read_coverage_distribution_help").click(function() {
                  $("#readCoverageDistributionModal").modal({
                    show: true
                  });
                });

                $("#fragment_help").click(function() {
                  $("#fragmentModal").modal({
                    show: true
                  });
                });

                $("#length_help").click(function() {
                  $("#lengthModal").modal({
                    show: true
                  });
                });

                $("#base_qual_help").click(function() {
                  $("#baseQualModal").modal({
                    show: true
                  });
                });

                $("#map_qual_help").click(function() {
                  $("#mapQualModal").modal({
                    show: true
                  });
                });

                // $("#read-coverage-distribution input[type='checkbox']").change(function(){
                //     if ($(this).parent().hasClass("checked"))
                //         window.readCoverageDistChart(hist2array(window.sampleStats.coverage_hist), { noOutliers : false })
                //     else
                //         window.readCoverageDistChart(hist2array(window.sampleStats.coverage_hist), { noOutliers : true })
                // });

                $("#scale-switch input[type='checkbox']").change(function(){
                  var selection = d3.select('#depth-distribution .chart')

                  if ( $(this).parent().hasClass("checked") )
                    window.readDepthChart.lineChart().y( window.readDepthChart.lineChart().y().exponent(0.5) );
                  else
                    window.readDepthChart.lineChart().y( window.readDepthChart.lineChart().y().exponent(1) );

                  window.readDepthChart(selection, { selected: getSelectedSeqId()});
                })

                $("#length-distribution input[type='checkbox']").change(function(){
                    var outliers = $(this).parent().hasClass("checked");

                    var dataId = $("#length-distribution .chart-chooser .selected").attr("data-id")
                    var h = window.sampleStats[dataId];
                    var d = Object.keys(h).map(function(k) { return  [+k, +h[k]] });
                    var selection = d3.select("#length-distribution-chart")
                    if (!outliers) d = iobio.viz.layout.outlier()(d);
                    selection.datum(d);
                    window.lengthChart(selection, {'outliers':outliers});
                });

                $("#url-input, #bai-url-input").keyup(function(event){
                    if(event.keyCode == 13){
                        $("#bam-url-go-button").click();
                    }
                });

                $('#url-input').focus(function() {
                    if (this.setSelectionRange) {
                        this.setSelectionRange(7, 7);
                    } else if (this.createTextRange) {
                        // IE
                        var range = this.createTextRange();
                        range.collapse(true);
                        range.moveStart('character', 7);
                        range.moveEnd('character', 7);
                        range.select();
                    }
                });

                document.getElementById('file').addEventListener('change', openBamFile, false);
                document.getElementById('bedfile').addEventListener('change', openBedFile, false);


                // initialize charts
                // get height width of histogram charts and set viewboxes
                var width = $("#read-coverage-distribution-chart").width();
                var height = $("#read-coverage-distribution-chart").height();
                var dists = document.getElementsByClassName("focus") // viewboxes need to be set at runtime to get accurate height/width
                for (var i=0; i < dists.length; i++ ) { dists[i].setAttribute('viewBox', "0 0 " + width + " " + height) }

                // get width for main panel depth chart
                var depthChartWidth = $("#depth-distribution .chart").width();
                var pieChooserPadding = 30;
                var pieChooserWidth = Math.min(+$("#piechooser").width(), +$("#piechooser").height() - 0) ;

                var color = d3.scale.category20b();

                var r = pieChooserWidth/2 - pieChooserPadding;

                window.draw = true;

                window.pieChooserChart = iobio.viz.pieChooser()
                    .radius(r)
                    .innerRadius(r*.5)
                    .padding(pieChooserPadding)
                    .transitionDuration(0)
                    .color( function(d,i) {
                      return color(d.data.name);
                    })
                    .on("click", function(d,i) {
                      setSelectedSeq(d.data.name);
                    })
                    .on("clickall", function(d,i) {
                      window.readDepthChart.trigger('click', 'all');
                    })
                    .tooltip( function(d) {
                      return d.data.name;
                    });


                // setup read coverage histogram chart
                window.readCoverageChart = iobio.viz.barViewer()
                  .xValue(function(d) { return d[0]; })
                  .yValue(function(d) { return d[1]; })
                  .wValue(function() { return 1; })
                  .height(height)
                  .width(width)
                  .margin({top: 5, right: 20, bottom: 20, left: 50})
                  .sizeRatio(0.8);
                window.readCoverageChart.yAxis().tickFormat(function(d) { return d*100 + '%'});

                window.lengthChart = iobio.viz.barViewer()
                  .xValue(function(d) { return d[0]; })
                  .yValue(function(d) { return d[1]; })
                  .wValue(function() { return 1; })
                  .height(height)
                  .width(width)
                  .margin({top: 5, right: 20, bottom: 20, left: 50})
                  .sizeRatio(0.8);
                window.lengthChart.xAxis().tickFormat(tickFormatter);
                window.lengthChart.yAxis().tickFormat(tickFormatter);

                // setup quality histogram chart
                window.qualityChart = iobio.viz.barViewer()
                    .xValue(function(d) { return d[0]; })
                    .yValue(function(d) { return d[1]; })
                    .wValue(function() { return 1; })
                    .height(height)
                    .width(width)
                    .margin({top: 5, right: 20, bottom: 20, left: 50})
                    .sizeRatio(0.8);
                window.qualityChart.xAxis().tickFormat(tickFormatter);
                window.qualityChart.yAxis().tickFormat(tickFormatter);

                // setup main panel read chart
                var yscale = d3.scale.pow().exponent(1);
                window.readDepthChart = iobio.viz.multiLine()
                  .nameValue(function(d) { return d.name; })
                  .dataValue(function(d) { return d.data; })
                  .xValue(function(d,i) { return d.pos; })
                  .yValue(function(d,i) { return d.depth; })
                  .wValue(function() { return 1; })
                  .transitionDuration(400)
                  .epsilonRate(0.1)
                  .margin({top: 10, right: 0, bottom: 30, left:0})
                  .height(150)
                  .color(function(d,i) {return color(d.name); })
                  .width(depthChartWidth)
                  .on('click', function(d) {
                    var name = d ? d.name : 'all';
                    setSelectedSeq(name);
                  })
                  .brush('brushend', function(b) {
                    var start = parseInt(b.extent()[0]), end = parseInt(b.extent()[1]);
                    if (start != end)
                      setSelectedSeq( window.readDepthChart.getSelected(), start, end );
                    else
                      setSelectedSeq( window.readDepthChart.getSelected() );
                  });

                window.readDepthChart.lineChart().y(yscale);

                // setup donut chart
                // window.sampleDonutChart = donutD3().radius(61).klass("sampleArc");
                window.sampleDonutChart = iobio.viz.pie()
                    .radius(61)
                    .innerRadius(50)
                    .color( function(d,i) { if (i==0) return '#2d8fc1'; else return 'rgba(45,143,193,0.2)'; });

                // hold onto stats
                window.sampleStats = undefined;

                // default sampling values
                window.samplingBinSize = 40000;
                window.binNumber = 20;
                window.binSize = 40000;
                window.sampleMultiplier = 1;
                window.sampleMultiplierLimit = 4;

                window.bam = undefined;
                window.sampling = getUrlParameter('sampling')

                // check if app is triggered from illumina
                var action = getUrlParameter('action')
                if (action == 'trigger') {
                  getIlluminaBamBaiUrls(function(bamUrl, baiUrl) {
                    window.bam = new Bam(bamUrl, { bai: baiUrl});
                    let region = parseRegionParam(getUrlParameter('region'));
                    goBam(region);
                  })
                }

                // check if app is triggered from illumina
                var source = getUrlParameter('source')

                if (source !== undefined) {
                  // if (window.location.hash != "") {
                  //   let newUrl = window.location.href.replace('#', '&')
                  //   window.location.href = newUrl;
                  // }
                  var project_id = getUrlParameter('project_id')

                  if(project_id === undefined) {
                    getOldHubIobioUrls(source, function(bamUrl, baiUrl) {
                      window.bam = new Bam(bamUrl, { bai: baiUrl});
                      let region = parseRegionParam(getUrlParameter('region'));
                      goBam(region);
                    })
                  } else {
                    getMosaicIobioUrls(source, function(bamUrl, baiUrl) {
                      window.bam = new Bam(bamUrl, { bai: baiUrl});
                      let region = parseRegionParam(getUrlParameter('region'));
                      goBam(region);
                    })
                  }
                }

                // check if url is present
                var bamUrl = decodeUrl(getUrlParameter('bam'))
                // check if url to bam file is supplied in url and sample it
                if (bamUrl != undefined) {
                    // remove https if present
                    // if (bamUrl.slice(0,5) == 'https')
                    //   bamUrl = 'http' + bamUrl.slice(5,bamUrl.length);
                    var baiUrl = decodeUrl(getUrlParameter('bai'));
                    if (baiUrl)
                        window.bam = new Bam( bamUrl, {bai: baiUrl} );
                    else
                        window.bam = new Bam( bamUrl );
                    var r = getUrlParameter('region');
                    let region = parseRegionParam(r);
                    goBam(region);
                }
            });

            function decodeUrl(url) {
              if (url && (url.slice(0,14) == 'https%3A%2F%2F' || url.slice(0,13) == 'http%3A%2F%2F'))
                return decodeURIComponent(url)
              else
                return url;
            }

            function parseRegionParam(regionParam) {
              var region = undefined;
              if (regionParam != undefined) {
                if (regionParam.split(":").length == 1)
                  region = { chr: regionParam.split(":")[0]}
                else
                  region = {
                    chr: regionParam.split(":")[0],
                    start: parseInt(regionParam.split(":")[1].split('-')[0]),
                    end: parseInt(regionParam.split(":")[1].split('-')[1])
                  };
              }
              return region;
            }

            function getUrlParameter(sParam) {
                // get params
                var sPageURL = window.location.search.substring(1);
                var sURLVariables = sPageURL.match(/(?:[^&"]+|"[^"]*")+/g) || [];
                // get hashes
                var sHashURL = window.location.hash.slice(1);
                var sHashVariables = sHashURL.match(/(?:[^&"]+|"[^"]*")+/g) || [];

                var variables = sURLVariables.concat(sHashVariables);

                for (var i = 0; i < variables.length; i++)
                {
                    var sParameterName = variables[i].match(/(?:[^="]+|"[^"]*")+/g);
                    if (sParameterName[0] == sParam)
                    {
                        return sParameterName[1];
                    }
                }
            }

            function removeBedFile() {
                $("#remove-bedfile-button").css('visibility', 'hidden');
                $("#default-bedfile-button").css('visibility', 'visible');
                $("#add-bedfile-button").css('visibility', 'visible');
                window.bed = undefined;
                goSampling({sequenceNames :  getSelectedSeqIds() });
            }

            function addDefaultBedFile() {
                var bedurl = '/20130108.exome.targets.bed';

                // clear brush on read coverage chart
        	    resetBrush();

                // hide add bed / show remove bed buttons
                $("#add-bedfile-button").css('visibility', 'hidden');
                $("#default-bedfile-button").css('visibility', 'hidden');
                $("#remove-bedfile-button").css('visibility', 'visible');

                // turn on sampling message and off svg
                // turn it on here b\c the bed file is so big it takes a while to download
                $("section#middle svg").css("display", "none");
                $(".samplingLoader").css("display", "block");


                // grab bed from url
                $.ajax({
                    // XDomainRequest protocol (IE 8 & 9) must be the same scheme as the calling page
                    url: bedurl,
                    dataType: 'text'
                }).done(function (data) {
                    data = data.replace(/chr/g, '');
                    window.bed = data;
                    goSampling({sequenceNames : getSelectedSeqIds() });
                });

            }

            function openBedFile(event) {

                if (event.target.files.length != 1) {
                   alert('must select a .bed file');
                   return;
                }

                // check file extension
                var fileType = /[^.]+$/.exec(event.target.files[0].name)[0];
                if (fileType != 'bed')  {
            	    alert('must select a .bed file');
            	    return;
        	    }
        	    // clear brush on read coverage chart
        	    resetBrush();

                // hide add bed / show remove bed buttons
                $("#add-bedfile-button").css('visibility', 'hidden');
                $("#default-bedfile-button").css('visibility', 'hidden');
                $("#remove-bedfile-button").css('visibility', 'visible')

	            // read bed file and store
	            var reader = new FileReader();
                reader.onload = function(theFile) {
                    window.bed = this.result;
                    goSampling({sequenceNames : getSelectedSeqIds() });
                }
                reader.readAsText(event.target.files[0])
              }

            function openBamFile(event) {

                if (event.target.files.length != 2) {
                   alert('must select both a .bam and .bai file');
                   return;
                }

                var fileType0 = /[^.]+$/.exec(event.target.files[0].name)[0];
                var fileType1 = /[^.]+$/.exec(event.target.files[1].name)[0];

                if (fileType0 == 'bam' && fileType1 == 'bai') {
            	    bamFile = event.target.files[0];
            	    baiFile = event.target.files[1];
            	 } else if (fileType1 == 'bam' && fileType0 == 'bai') {
            	    bamFile = event.target.files[1];
            	    baiFile = event.target.files[0];
            	 } else {
            	    alert('must select both a .bam and .bai file');
            	 }

                 window.bam = new Bam( bamFile, { bai: baiFile });
                 goBam();
              }

             function goBam(region) {
                 $("#selectData").css("display", "none");
                 $("#showData").css("visibility", "visible");

                 // get read depth
                 window.bam.estimateBaiReadDepth(function(id,points, done){
                     // setup first time and sample

                     if ( Object.keys(window.bam.readDepth).length == 1) {
                         // turn off read depth loading msg
                         $("#readDepthLoadingMsg").css("display", "none");
                         // turn on sampling message
                         $(".samplingLoader").css("display", "block");

                     }

                     var allPoints = Object.keys(window.bam.readDepth)
                      .filter(function(key) {
                        if (filterRef(key))
                          return false
                        if (window.bam.readDepth[key].length > 0)
                          return  true
                      })
                      .map(function(key) {
                        return {"name" : key, "data" : window.bam.readDepth[key] }
                     })

                    draw = true;

                    var numRefs = (d3.select('#depth-distribution .chart').datum() || []).length;

                    var selection = d3.select('#depth-distribution .chart').datum(allPoints);

                    var pie = d3.layout.pie()
                       .sort(null)
                       .value(function(d,i) {return d.data.length });

                    var pieSelection = d3.select('#piechooser').datum( pie(allPoints) )

                    if (allPoints.length > 50) {
                      $('#piechooser svg').css('visibility', 'hidden');
                      $('.too-many-refs').css('display', 'block')
                      draw = false;
                    }

                    if (region && region.chr != 'all') {
                      if(draw && (allPoints.length > numRefs)) window.readDepthChart(selection, {selected:region.chr, noLine:!done});
                    }
                    else {
                      if(draw && (allPoints.length > numRefs)) window.readDepthChart(selection, {noLine:!done});
                    }

                    $('#reference-select')
                      .append($("<option></option>")
                        .attr("value", id)
                        .text(id));

                     var start = region ? region.start : undefined;
                     var end = region ? region.end : undefined;
                     // if (region && region.chr == id ) {
                     //     window.pieChooserChart(pieSelection);
                     //     setSelectedSeq( id, start, end);
                     //     //window.readDepthChart(selection, {selected:id});
                     // }

                     if ( done ) {
                        window.pieChooserChart.on('end', function() {
                          if (!region || (region && region.chr == 'all'))
                            setSelectedSeq( 'all' , start, end );
                          else
                            setSelectedSeq( region.chr, start, end);
                        })
                        window.pieChooserChart(pieSelection);
                     }

                     var totalPoints = allPoints.reduce(function(acc,val) { return acc + val.data.length  },0)
                     if (done && totalPoints <= 1) {
                        $('#not_enough_data').css('display','block');
                     }
                 });

             }

            function resetRegionStats() {
                window.regionStats = {'Total reads' : {number : 0}};
            }

            function resetBrush() {
                var brush = window.readDepthChart.brush();
                var g = d3.selectAll("#depth-distribution .iobio-brush")
                brush.clear();
                brush(g);
                // remove brush region from url
                setUrlRegion({chr:getSelectedSeqId()});
                // no need to trigger event? since setSelected sq will do it
                //brush.event(g);
            }


            function filterRef(ref) {
              return (
                ref.substr(0,4) == 'GL00' ||
                ref.substr(0,6).toLowerCase() == "hs37d5" ||
                ref.includes('GL00') ||
                ref.includes('KI2707') ||
                ref.split("_").slice(-1)[0] == "alt" ||
                ref.split("_").slice(-1)[0] == "decoy" ||
                ref.includes('chrUn') ||
                ref.substr(0,4) == 'HLA-'
              )
            }


            function updateTotalReads(totalReads) {
                // update total reads
                var reads = shortenNumber( totalReads );
                $("#total-reads>#value").html( reads[0] );
                $("#total-reads>#base>#number").html( reads[1] || "" );
            }

            function setSelectedSeq(selected, start, end) {
                if (selected == 'all') {
                  var seqDataIds = Object.keys(window.bam.readDepth)
                        .filter(function(key) {
                            if (filterRef(key))
                              return false
                            if (window.bam.readDepth[key].length > 0)
                              return  true
                          })
                  // setTimeout(function() {
                    window.pieChooserChart.clickAllSlices(d3.selectAll('#piechooser .arc')[0]);
                  // }, 2000);
                }
                else {
                  var seqDataIds = [selected];
                  $('#piechooser .arc').each(function(i,d) {
                    if (d3.select(d).datum().data.name == selected) {
                      // setTimeout(function() {window.pieChooserChart.clickSlice(i);}, 2000);
                      window.pieChooserChart.clickSlice(i)
                    }

                  });

                }

                // window.readDepthChart.trigger('click', selected);
                if(draw) window.readDepthChart.setSelected(selected);

                $("#reference-select").val(selected);

                // reset brush
                resetBrush();
                setUrlRegion({chr:selected, 'start':start, 'end':end });
               // start sampling
               if(start!= undefined && end!=undefined) {
                 goSampling({ sequenceNames:seqDataIds, 'start':start, 'end':end });
                 setTimeout(function() { setBrush(start,end)}, 200);
               } else {
                 goSampling({ sequenceNames:seqDataIds});
               }
            }

            function setBrush(start,end) {
              var brush = window.readDepthChart.brush();
                 // // set brush region
              d3.select("#depth-distribution .iobio-brush").call(brush.extent([start,end]));
            }

            function resetBrush() { setBrush(0,0); }

            function setUrlRegion(region) {
              if (window.bam.sourceType == 'url' && region != undefined) {
                if (region.start != undefined && region.end != undefined) {
                  var regionStr = region.chr + ':' + region.start + '-' + region.end;
                } else {
                  var regionStr = region.chr;
                }
                var extraParams = '';

                if (window.sampling) extraParams += '&sampling=' + window.sampling

                if (getUrlParameter('source') !== undefined) {
                  const source = getUrlParameter('source');
                  window.history.pushState({'index.html' : 'bar'},null,"?source="+source+"&sample_uuid=" + getUrlParameter('sample_uuid') + "&region=" + regionStr + extraParams);

                } else if (window.bam.baiUri != undefined) {
                  window.history.pushState({'index.html' : 'bar'},null,"?bam=" + encodeURIComponent(window.bam.bamUri) + "&bai=" + encodeURIComponent(window.bam.baiUri) + "&region=" + regionStr + extraParams);

                } else {
                  window.history.pushState({'index.html' : 'bar'},null,"?bam=" + encodeURIComponent(window.bam.bamUri) + "&region=" + regionStr + extraParams);
                }
              }
            }

            function goSampling(options) {
                // add default options
                options = $.extend({
                    exomeSampling : 'checked' == $("#depth-distribution input").attr("checked"),
                    bed : window.bed,
                    onEnd:function(){NProgress.done();}
                },options);

                // turn on sampling message and off svg
                $("section#middle svg").css("display", "none");
                $(".samplingLoader").css("display", "block");
                updateTotalReads(0);
                NProgress.start();
                NProgress.set(0);
                // update selected stats
                window.bam.sampleStats( function(data){
                    // turn off sampling message
                    $(".samplingLoader").css("display", "none");
                    $("section#middle svg").css("display", "block");
                    window.sampleStats = data;
                    // update progress bar
                    if (options.start != null && options.end != null) {
                        var length = options.end - options.start;
                        var percentDone = Math.max(Math.round( ((data.last_read_position-options.start) / length) * 100) / 100,0);
                    } else {
                        var length = window.bam.header.sq.reduce(function(prev,curr) { if (prev)return prev; if (curr.name == options.sequenceNames[0] )return curr; }, false).end;
                        var percentDone = Math.round( (data.last_read_position / length) * 100) / 100;
                    }

                    if (NProgress.status < percentDone) NProgress.set(percentDone);
                    // update charts
                    updatePercentCharts(data, window.sampleDonutChart)
                    updateTotalReads(data.total_reads);
                    updateHistogramCharts(data, undefined, "sampleBar")
                },options);
            }

            function sampleMore() {
                if (window.sampleMultiplier >= window.sampleMultiplierLimit) { alert("You've reached the sampling limit"); return;}
                window.sampleMultiplier += 1;
                var options = {
                    sequenceNames : getSelectedSeqIds() ,
                    binNumber : window.binNumber + parseInt(window.binNumber/4 * window.sampleMultiplier),
                    binSize : window.binSize + parseInt(window.binSize/4 * window.sampleMultiplier)
                }
                if (window.readDepthChart.brush().extent().length != 0 && window.readDepthChart.brush().extent().toString() != "0,0") {
                    options.start = parseInt(window.readDepthChart.brush().extent()[0]);
                    options.end = parseInt(window.readDepthChart.brush().extent()[1]);
                }
                goSampling(options);
            }

            function getSelectedSeqId() {
              return window.readDepthChart.getSelected();
            }

            function getSelectedSeqIds() {
              var selected = window.readDepthChart.getSelected();
              if (selected == 'all') {
                return Object.keys(window.bam.readDepth)
                      .filter(function(key) {
                        if (filterRef(key))
                          return false
                        if (window.bam.readDepth[key].length > 0)
                          return  true
                      })
              } else
                return [selected];
            }

            function updatePercentCharts(stats, donutChart) {
                var pie = d3.layout.pie()
                   .sort(null);
                var value = undefined;

                var selected = getSelectedSeqId();
                var unmappedReads, mappedReads;
                if (selected == 'all') {
                  if (window.bam.readDepth[Object.keys(window.bam.readDepth)[0]].mapped != undefined) {
                    mappedReads = unmappedReads = 0;
                    for ( var id  in window.bam.readDepth) {
                      mappedReads += window.bam.readDepth[id].mapped;
                      unmappedReads += window.bam.readDepth[id].unmapped;
                    }
                    unmappedReads = window.bam.n_no_coor;
                  }
                }
                else {
                  mappedReads = window.bam.readDepth[selected].mapped;
                  unmappedReads = window.bam.readDepth[selected].unmapped;
                }
                var showMappedDataFromIndex = false;
                var brushRange = window.readDepthChart.brush().extent();
                if ( (brushRange == undefined || brushRange.toString() == '0,0' ) && mappedReads != undefined && unmappedReads != undefined) {
                  showMappedDataFromIndex = true;
                  d3.select("#mapped_reads").selectAll('path')
                    .attr('fill', function(d,i) { return i==0 ? 'rgb(9,176,135)' : 'rgba(9,176,135,0.5)' });
                  d3.select('.percent .from-index').style('visibility', 'visible')
                }
                else
                  d3.select('.percent .from-index').style('visibility', 'hidden')

                //update percent charts
                var keys = ['mapped_reads', "proper_pairs", "forward_strands", "singletons", "both_mates_mapped", "duplicates"]
                // var colors = ["rgb(45,143,193)", 'rgb(231, 76, 60)', "rgb(243, 156, 18)", "rgb(155, 89, 182)", "rgb(46, 204, 113)", "rgb(241, 196, 15)"];
                keys.forEach( function(key,i) {
                    var stat = stats[key];
                    if (key == 'mapped_reads' && showMappedDataFromIndex)  {
                      var data = [ mappedReads, unmappedReads ];
                    }
                    else {
                      if (stats['total_reads'] == 0)
                        var data = [ 0, 100 ];
                      else
                        var data = [ stat, stats['total_reads'] - stat ];
                    }

                    var selection = d3.select('#'+key  + ' .chart').datum(pie(data));
                    donutChart(selection);
                });

            }

            function updateHistogramCharts(histograms, otherMinMax, klass) {

                // check if coverage is zero
                if (Object.keys(histograms.coverage_hist).length == 0) histograms.coverage_hist[0] = '1.0';
                // update read coverage histogram
                var d = Object.keys(histograms.coverage_hist).filter(function(i){return histograms.coverage_hist[i] != "0"}).map(function(k) { return [+k, +histograms.coverage_hist[k]] });
                var selection = d3.select("#read-coverage-distribution-chart").datum(d);
                window.readCoverageChart(selection);
                // if (histograms.coverage_hist[0] > 0.65) {
                //     // most likely exome
                //     // exclude <5 values b\c they are not informative dominating the chart
                //     var min = 5;
                //     var max = window.readCoverageChart.globalChart().x().domain()[1];
                //     window.readCoverageChart.setBrush([min, max]);
                // }

                // update read length distribution
                if ($("#length-distribution .selected").attr("data-id") == "frag_hist")
                    var d = Object.keys(histograms.frag_hist).filter(function(i){return histograms.frag_hist[i] != "0"}).map(function(k) { return  [+k, +histograms.frag_hist[k]] });
                else
                    var d = Object.keys(histograms.length_hist).map(function(k) { return  [+k, +histograms.length_hist[k]] });
                // remove outliers if outliers checkbox isn't explicity checked
                var outliers = $("#length-distribution .checkbox").hasClass("checked");
                if (!outliers) d = iobio.viz.layout.outlier()(d);
                var selection = d3.select("#length-distribution-chart").datum(d);
                window.lengthChart( selection );
                // window.lengthChart( selection, {'outliers':outliers} );

                // update map quality distribution
                if ($("#mapping-quality-distribution .selected").attr("data-id") == "mapq_hist")
                    var d = Object.keys(histograms.mapq_hist).map(function(k) { return  [+k, +histograms.mapq_hist[k]] });
                else
                    var d = Object.keys(histograms.baseq_hist).map(function(k) { return  [+k, +histograms.baseq_hist[k]] });
                var selection = d3.select("#mapping-quality-distribution-chart").datum(d);
                window.qualityChart(selection);
            }

            function getIlluminaAccessToken(callback) {
                var basespace = 'nv-dev-new.iobio.io/basespaceauth/';
                var authorization_code = getUrlParameter('authorization_code');
                var cmd = new iobio.cmd(basespace, [authorization_code]);

                cmd.on('data', function(d) {
                    var accessToken = JSON.parse(d).access_token
                    callback(accessToken);
                })
                    .on('error', function(e) { /*console.error(e)*/ ;})
                    .run()
            }

            function getOldHubIobioUrls(source, callback) {
              let api = decodeURIComponent(source) + "/apiv1";

              let access_token = getUrlParameter('access_token');
              let sample_uuid = getUrlParameter('sample_uuid');
              let token_type = getUrlParameter('token_type');

              if (access_token !== undefined) {
                localStorage.setItem('hub-iobio-tkn', token_type + ' ' + access_token);
              }

              if (localStorage.getItem('hub-iobio-tkn')) {

              // Get VCF File
                getFilesForSample(sample_uuid).done(data => {
                  bam = data.data.filter(f => f.type == 'bam')[0];
                  bai = data.data.filter(f => f.type == 'bai')[0];

                  // Get Signed Url
                  getSignedUrlForFile(bam).done(bamUrlData => {
                    bamUrl = bamUrlData.url;
                    getSignedUrlForFile(bai).done(baiUrlData => {
                      baiUrl = baiUrlData.url;
                      callback(bamUrl, baiUrl);
                    })
                  })
                })
              } else {
                window.location.href = buildOauthLink();

              }

              function getFilesForSample(sample_uuid) {
                return $.ajax({
                  url: api + '/samples/' + sample_uuid + '/files',
                  type: 'GET',
                  contentType: 'application/json',
                  headers: {
                    'Authorization': localStorage.getItem('hub-iobio-tkn')
                  }
                }).fail(function(xhr,status,error) {
                  let = link = buildOauthLink();
                  $('#warning-authorize')
                    .append('Your access to hub.iobio has expired. Please click <a href='+link+'>here</a> to renew your access.')
                    .css('display', 'block');
                });
              }

              function getSignedUrlForFile (file) {
                return $.ajax({
                  url: api + '/files/' + file.uuid + '/url',
                  type: 'GET',
                  contentType: 'application/json',
                  headers: {
                    'Authorization': localStorage.getItem('hub-iobio-tkn')
                  }
                });
              }

            }

            function getMosaicIobioUrls(source, callback) {
              let api = decodeURIComponent(source) + "/apiv1";

              let project_id = getUrlParameter('project_id');
              let access_token = getUrlParameter('access_token');
              let sample_id = getUrlParameter('sample_id');
              let token_type = getUrlParameter('token_type');

              if (access_token !== undefined) {
                localStorage.setItem('hub-iobio-tkn', token_type + ' ' + access_token);
              }

              if (localStorage.getItem('hub-iobio-tkn')) {

              // Get VCF File
                getFilesForSample(sample_id).done(data => {
                  bam = data.data.filter(f => f.type == 'bam')[0];
                  bai = data.data.filter(f => f.type == 'bai')[0];

                  // Get Signed Url
                  getSignedUrlForFile(project_id, bam).done(bamUrlData => {
                    bamUrl = bamUrlData.url;
                    getSignedUrlForFile(project_id, bai).done(baiUrlData => {
                      baiUrl = baiUrlData.url;
                      callback(bamUrl, baiUrl);
                    })
                  })
                })
              } else {
                window.location.href = buildOauthLink();

              }

              function getFilesForSample(sample_id) {
                return $.ajax({
                  url: api + '/samples/' + sample_id + '/files',
                  type: 'GET',
                  contentType: 'application/json',
                  headers: {
                    'Authorization': localStorage.getItem('hub-iobio-tkn')
                  }
                }).fail(function(xhr,status,error) {
                  let = link = buildOauthLink();
                  $('#warning-authorize')
                    .append('Your access to hub.iobio has expired. Please click <a href='+link+'>here</a> to renew your access.')
                    .css('display', 'block');
                });
              }

              function getSignedUrlForFile (project_id, file) {
                return $.ajax({
                  // url: api + '/files/' + file.id + '/url',
                  url: api + '/projects/' + project_id + '/files/' + file.id + '/url',
                  type: 'GET',
                  contentType: 'application/json',
                  headers: {
                    'Authorization': localStorage.getItem('hub-iobio-tkn')
                  }
                });
              }

            }

            function buildOauthLink() {
              // var client_id = apiConfig['hub_client'].id;
              var hubApi = decodeURIComponent(getUrlParameter('source'));
              var client_id = 'u3oRvGom';
              // var queryString = qs.stringify(queryParams, { arrayFormat: 'brackets', addQueryPrefix: true });
              var redirect_uri = encodeURIComponent(window.location.href);
              return `${hubApi}/oauth/authorization?response_type=token&client_id=${client_id}&redirect_uri=${redirect_uri}`
            }

            function getIlluminaBamBaiUrls(callback) {
                var appSessionHref = getUrlParameter('appsessionuri'),
                    apiUrl = "https://api.basespace.illumina.com/";

                getIlluminaAccessToken(function(accessToken) {
                    $.ajax({
                       url: apiUrl + appSessionHref,
                       data: {access_token: accessToken},
                       type: 'GET',
                       contentType: "application/json",
                       dataType: 'json',
                       error: function(error) {
                          console.error(error);
                       }

                    }).done(function(sessionRes) {
                        var bamHref, baiHref;

                        sessionRes.Response.References.forEach(function(ref) {
                            if (ref.Type == 'File' && ref.Rel == 'Input') {
                                if (ref.Content.Name.slice(-3) == 'bam')
                                    bamHref = ref.Content.HrefContent;
                                else if (ref.Content.Name.slice(-3) == 'bai')
                                    baiHref = ref.Content.HrefContent;
                            }
                        })

                        // get bam S3 location
                        $.ajax({
                            url: apiUrl + bamHref,
                            data: {access_token: accessToken, redirect: 'meta'},
                            type: 'GET',
                            contentType: "application/json",
                            dataType: 'json',
                            error: function(error) {
                                console.error(error);
                            }
                        }).done(function(bamRes) {

                            var bamUrl = '"' + bamRes.Response.HrefContent + '"' ;

                            // get bai s3 location
                            $.ajax({
                                url: apiUrl + baiHref,
                                data: {access_token: accessToken, redirect: 'meta'},
                                type: 'GET',
                                contentType: "application/json",
                                dataType: 'json',
                                error: function(error) {
                                    console.error(error);
                                }
                            }).done(function(baiRes) {
                                var baiUrl = '"' + baiRes.Response.HrefContent + '"';
                                callback(bamUrl, baiUrl);
                            })
                        })
                    })
                })
            }


            function toggleChart(elem, chartId) {
                if ($(elem).hasClass("selected")) return;
                // toggle selected
                var pair = [elem, $(elem).siblings()[0]];
                $(pair).toggleClass('selected');

                // redraw chart
                var dataId = elem.getAttribute("data-id")
                // var outlier = elem.getAttribute('data-outlier') === 'true' || elem.getAttribute('data-outlier') == null;
                var h = window.sampleStats[dataId];
                var d = Object.keys(h).map(function(k) { return  [+k, +h[k]] });
                var chartDiv = $(elem).parent().parent().parent();
                var selection = d3.select(chartDiv.find('.chart')[0])
                if ( chartDiv.find(".selected").attr("data-id") == "frag_hist" ) {
                    var outlier = chartDiv.find('.checkbox').hasClass('checked')
                    if (!outlier) d = iobio.viz.layout.outlier()(d);
                }
                selection.datum(d);
                window[chartId](selection);
            }

            function shortenNumber(num) {
                if(num.toString().length <= 3)
                    return [num];
                else if (num.toString().length <= 6)
                    return [Math.round(num/1000), "thousand"];
                else
                    return [Math.round(num/1000000), "million"];
            }

            // function hist2array(hist) {
            //     var a = [];
            //     for ( var i in hist) {
            //        a.push([ i,hist[i] ]);
            //         // var position = i;
            //         //                     var instances = hist[i];
            //         // for ( var j=0; j < instances; j++)
            //         //                         a.push( parseInt(position) );
            //     }
            //     return a;
            // }

            function displayBamUrlBox() {
                $('#bam-url').css('visibility','visible');
                $('#info').css('visibility', 'hidden');
                $("#bam-url").children("input").first().focus();
            }

            function openBamUrl() {
                var url = $("#url-input").val();
                var baiUrl = $("#bai-url-input").val()
                // remove https if present
                // if (url.slice(0,5) == 'https')
                //   url = 'http' + url.slice(5,url.length);


                if( baiUrl != '' ) {
                  window.history.pushState({'index.html' : 'bar'},null,"?bam=" + encodeURIComponent(url) + "&bai=" + encodeURIComponent(baiUrl));
                  window.bam = new Bam(url, { bai: baiUrl});
                } else {
                  window.history.pushState({'index.html' : 'bar'},null,"?bam=" + encodeURIComponent(url));
                  window.bam = new Bam( url );
                }
                goBam();
            }

            function tickFormatter (d) {
              if ((d / 1000000) >= 1)
                d = d / 1000000 + "M";
              else if ((d / 1000) >= 1)
                d = d / 1000 + "K";
              return d;
           }

        </script>
    </head>
    <body>
        <div id="warning-authorize">
          Test
        </div>
        <header>
            <a href="http://bam.iobio.io">bam<span style="color:rgb(200,200,200)">.iobio.io</span></a>
            <div style="float:right" class="iobio-project" style="padding-top:20px">
                <a href="http://iobio.io" style="margin: 0px 5px 0px 5px;">an iobio project</a>
            </div>
        </header>
        <div id='maintenance_mode' style="display: none; text-align: center; width:80%; font-family: arial; font-size: 30px; margin: 100px auto;">
          We are currently undergoing maintenance please check back in a few minutes
        </div>

        <div class="beta-info">
          <a class="btn btn-raised beta-link"  href="http://newbam.iobio.io">
            <svg class="beta-icon" fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
                <path d="M0 0h24v24H0z" fill="none"/>
                <path d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"/>
            </svg>
            Try our beta release
          </a>
        </div>

        <div id="content">
          <div id="selectData">
              <div style="width:700px; margin-left:auto; margin-right:auto; margin-top: 100px">
                  <div style="margin-left:auto;margin-right:auto;font-size: 28px; color: rgb(110,110,110); margin-bottom:70px">
                      examine your sequence alignment file in seconds
                  </div>

                  <div class="file" onClick="$('#bam-url').css('visibility', 'hidden')">
                      <input type="file" name="files[]" id="file"  multiple />
                      <label class="file-button" for="file" >choose bam file</label>
                  </div>
                  <div class="file-button" style="text-align:center" onClick="displayBamUrlBox()">choose bam url</div>
                  <div style="clear:both"></div>
                  <div id="info">
                      <ul>
                              <li><a href="http://www.nature.com/nmeth/journal/v11/n12/full/nmeth.3174.html">Publication</a></li>
                              <li><a href="help.html">File Requirements</a></li>
                              <li><a href="license.html">License</a></li>
                              <li><a href="images/browserCompatability.png">Compatible Browsers</a></li>
                      </ul>
                  </div>
              </div>
              <div id='bam-url' style="margin-top:-21px;width:700px;margin-left:auto;margin-right:auto;visibility:hidden" class="arrow_box">
                  <input id="url-input" value="http://s3.amazonaws.com/iobio/NA12878/NA12878.autsome.bam" ></input>
                  <input id="bai-url-input" placeholder="BAI URL (optional)" ></input>
                  <button id="bam-url-go-button" onClick="openBamUrl()">Go</button>
              </div>
              <div style="margin: 50px auto 0px auto; font-size: 23px; width: 600px; text-align:center;  color: rgb(110,110,110)">
                  <div>for variant files check out <a href="http://vcf.iobio.io">vcf.iobio</a></div>
              </div>

              <div id="footer">
                  <div class="logos">
                      <div style="position:absolute; left: 100px; top:4px; font-size:50px;font-family:'Overlock SC', cursive;">Marthlab</div>
                      <img src="images/ustar-ucgd-logo.jpg" style="height:60px;"/>
                      <a href="http://www.genetics.utah.edu/"><img src="images/genetics_mainlogo3_lrg.png" style="height:50px;position:absolute;right:0px;top:7px"/></a>
                  </div>
              </div>
          </div>
          <div id="showData">
              <section id="top">
                  <div id="piechooser" class="panel">
                    <select onchange='setSelectedSeq(this.value);' id="reference-select">

                      <option value="all">all</option>
                    </select>
                  </div>
                  <div id="depth-distribution" class="panel">

                      <div class="title">
                        <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="genome_coverage_help" title="Coverage across the genome" style="font-size:0.45em;"></span>
                        Read Coverage
                      </div>
                      <div class="hint">(drag to select region)</div>
                      <!-- <label class="checkbox" for="checkbox2" style="position:absolute;right:15px;top:21px;cursor:pointer" title="Turn on Exome Sampling - use the read coverage data to sample only regions with non-zero read depth">
                          <input type="checkbox"value="" class="outlier" data-toggle="checkbox" >
                          Exome
                      </label> -->
                      <input type="file" name="files[]" id="bedfile"  multiple />
                      <div id="remove-bedfile-button" class="bedfile-button" onClick="removeBedFile()" style="visibility:hidden">Remove Bed</div>
                      <div id="default-bedfile-button" class="bedfile-button" onClick="addDefaultBedFile()" title="1000G human exome targets file " style="right:110px">GRCh37 exonic regions</div>
                      <label id="add-bedfile-button" class="bedfile-button" for="bedfile" title="Add Bed format capture target definition file">Custom Bed</label>

                      <!-- log toggle -->
                      <label id="scale-switch" class="checkbox">
                              <input type="checkbox"value=""  class='power-scale' data-toggle="checkbox" >
                              Power Scale
                      </label>

                      <div id="readDepthLoadingMsg" style="font-size:50px;margin-top:30px;color:#2687BE">Initializing data <img style="height:18px" src="images/loading_dots.gif"/></div>
                      <div class='warning' id="not_enough_data">Bam file is too small to read coverage information</div>
                      <div class="warning too-many-refs">Too many references to display. Use the dropdown to the left to select the reference</div>
                      <!-- <div id="test-chart" style="width:100%"></div> -->
                      <div class='chart' style="width:98%; height:60%"></div>
                      <!-- <ul id="sequences"></ul> -->
                  </div>

                  <!-- Read coverage modal -->
                  <div id="genomeCoverageModal" class="modal fade" role="dialog">
                    <div class="modal-dialog" style="width:700px;">
                      <div class="modal-content">
                        <div class="modal-header">
                          <button type="button" class="close" data-dismiss="modal">&times;</button>
                          <h4 class="modal-title">Read coverage</h4>
                        </div>
                        <div class="modal-body" style="overflow:auto;">
                          <div>
                            The read coverage shows how the read coverage varies across the entire genome. The coloured numbers beneath represent chromosomes in the reference genome used and can be selected to view the read coverage in an individual chromosome. Selecting a different chromosome will cause all other metrics in bam.iobio to be recalculated based on reads sampled from that chromosome only. Once a chromosome is selected, you can also focus on a smaller region by dragging over the region of interest; again, all other metrics will then be recalculated for that region only.
                          </div>
                        </div>
                        <div class="modal-footer">
                          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                        </div>
                      </div>
                    </div>
                  </div>

                  <div id="total-reads" class="panel">
                      <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="sampled_help" title="Number of reads sampled"></span>
                      <div class="title">
                        Reads Sampled
                      </div>
                      <div id="value" style="font-size:5.5vw; color:#2687BE; text-align:center;padding-left:8px;">0</div>
                      <div id="base" style="font-size:1.4vw;padding-left:28px; letter-spacing:3px; color:#2687BE; text-align:center;">
                          <span id="number"></span>
                          <img title="Sample More" style="width:20px;margin-bottom:-3px;margin-left:-7px;cursor:pointer" onClick="sampleMore()"  src="images/more_sampling.png"></img>
                      </div>

                      <!-- Sampled reads modal -->
                      <div id="sampledModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Sampled reads</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                Bam.iobio does not read the entire bam file, rather, it samples reads from across the entire genome. The number of reads that have been sampled are shown here, and should be at least in the tens of thousands to have confidence in the statistics. Click the arrow beneath the displayed number to increase the number of sampled reads.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                  </div>
              </section>
              <section id="middle">
                  <div id="percents">

                      <!-- Mapped reads chart -->
                      <div id="mapped_reads" class="percent panel">Mapped Reads
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="mapped_reads_help" title="Expect a value >90%"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                          <div class='from-index'>* full data available in index</div>
                      </div>

                      <!-- Sampled reads modal -->
                      <div id="sampledModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Sampled reads</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                The mapped reads chart shows how many of the reads in the sample were successfully mapped to the reference genome. Genetic variation, in particular structural variants, ensure that every sequenced sample is genetically different to the reference genome it was aligned to. If the sample differs only in a small number of single base pair changes (e.g. SNVs), the read will still likely map to the reference, but, for more significant variation, the read can fail to be placed. Therefore, it is not expected that the mapped reads rate will hit 100%, but it is expected to be high (usually >90%).
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Mapped reads modal -->
                      <div id="mappedReadsModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Mapped reads</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                The mapped reads chart shows how many of the reads in the sample were successfully mapped to the reference genome. Genetic variation, in particular structural variants, ensure that every sequenced sample is genetically different to the reference genome it was aligned to. If the sample differs only in a small number of single base pair changes (e.g. SNVs), the read will still likely map to the reference, but, for more significant variation, the read can fail to be placed. Therefore, it is not expected that the mapped reads rate will hit 100%, but it is expected to be high (usually >90%).
                              </div>
                              <div class="row vertical-align">
                                  <div class="col-xs-4">
                                    <img title="Acceptable mapped reads rate" style="width:100%;"  src="images/mapped_reads_high.png"></img>
                                  </div>
                                  <div class="col-xs-8">
                                    This is an example of a human, whole exome. In this case, 99.7% of the sampled reads map to the reference, corresponding to 93,385 actual reads. It is important to note that when the wheel is blue, only reads that have been assigned to a reference sequence are included. This means that the 0.3% of reads that are unmapped have a mate pair that successfully maps to the reference genome.
                                  </div>
                              </div>

                              <div class="row vertical-align">
                                  <div class="col-xs-4">
                                    <img title="Acceptable mapped reads rate" style="width:100%;" src="images/mapped_reads_high_green.png"></img>
                                  </div>
                                  <div class="col-xs-8">
                                    For the case that both mates from paired end sequencing are unmapped, they appear at the end of the BAM file. Usually, the number of such unmapped reads can be obtained from the index file. When this is possible, the wheel will appear in green, as shown for this whole genome sample.
                                  </div>
                              </div>

                              <div class="row vertical-align">
                                  <div class="col-xs-4">
                                    <img title="Acceptable mapped reads rate" style="width:100%;" src="images/mapped_reads_low.png"></img>
                                  </div>
                                  <div class="col-xs-8">
                                    If the rate of mapped reads is low (usually below 90%), questions need to be asked about the sample to understand why so many reads are unmapped. The last example only has 71.5% of reads mapping to the reference genome for a whole genome sample. This was caused as the sample was contaminated with a significant amount of bacterial DNA; the DNA sample was obtained from a saliva sample, rather than a blood draw.
                                  </div>
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Strands chart -->
                      <div id="forward_strands" class="percent panel">Forward Strand
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="strands_help" title="Expect a value of ~50%"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                      </div>

                      <!-- Strands modal -->
                      <div id="strandsModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Forward strand</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                The forward strand chart shows the fraction of reads that map to the forward DNA strand. The general expectation is that the DNA library preparation step will generate DNA from the forward and reverse strands in equal amounts. After mapping the reads to the reference genome, approximately 50% of the reads will consequently map to the forward strand. If the observed rate is significantly different to 50%, this may be indicative of problems with the library preparation step.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Proper pairs chart -->
                      <div id="proper_pairs" class="percent panel">Proper Pairs
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="proper_pairs_help" title="Expect a value >90%"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                      </div>

                      <!-- Proper pairs modal -->
                      <div id="properPairsModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Proper Pairs</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                A fragment consisting of two <i>mates</i> is called a proper pair if both <i>mates</i> map to the reference genome in a manner consistent with expectations. In particular, if the DNA library consists of fragments ~500 base pairs in length, and 100 base pair reads are sequenced from either end, the expectation would be that the two reads map to the reference genome separated by ~300 base pairs. If the sequenced sample contains large structural variants, e.g. a large insertion, reads mapping with a large separation would be a signal for this variant, and the reads would not be proper pairs. Based on the sequencing technology, there is also an expectation on the orientation of each read in the fragment.
                              </div>
                              <br>
                              <div>
                                <i><strong>When calculating the proper pair rate, pairs where both mates are unmapped are not included in the analysis.</strong></i> As a consequence, the rate of proper pairs is expected to be well over 90%; even if the mapping rate itself is low as a result of bacterial contamination, for example.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Singletons chart -->
                      <div id="singletons" class="percent panel">Singletons
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="singletons_help" title="Expect a value <1%"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                      </div>

                      <!-- Singletons modal -->
                      <div id="singletonsModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Singletons</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                When working with paired-end sequencing, each DNA fragment is sequenced from both ends, creating two <i>mates</i> for each pair. If one <i>mate</i> in the pair successfully maps to the reference genome, but the other is unmapped, the mapped mate is a <i>singleton</i>. One way in which a singleton could occur would be if the sample has a large insertion compared with the reference genome; one <i>mate</i> can fall in sequence flanking the insertion and will be mapped, but the other falls in the inserted sequence and so cannot map to the reference genome. There are unlikely to many such structural variants in the sample, or sequencing errors that would could cause a read to not be able to map. Consequently, the singleton rate is expected to be very low (<1%).
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Both mates mapped chart -->
                      <div id="both_mates_mapped" class="percent panel">Both Mates Mapped
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="both_mapped_help" title="Expect a value >90%"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                      </div>

                      <!-- Both mates mapped modal -->
                      <div id="bothMappedModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Both mates mapped</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                When working with paired-end sequencing, each DNA fragment is sequenced from both ends, creating two <i>mates</i> for each pair. This chart shows the fraction of reads in pairs where both of the <i>mates</i> successfully map to the reference genome. <i><strong>When calculating this metric, pairs where both mates are unmapped are not included.</strong></i>.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Duplicates chart -->
                      <div id="duplicates" class="percent panel">Duplicates
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="duplicates_help" title="Value depends on depth"></span>
                          <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                          <div class='chart'></div>
                      </div>

                      <!-- Duplicates modal -->
                      <div id="duplicatesModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Duplicates</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                PCR duplicates are two (or more) reads that originate from the same DNA fragment. When sequencing data is analysed, it is assumed that each observation (i.e. each read) is independent; an assumption that fails in the presence of duplicate reads. Typically, algorithms look for reads that map to the same genomic coordinate, and whose mates also map to identical genomic coordinates. It is important to note that as the sequencing depth increases, more reads are sampled from the DNA library, and consequently it is increasingly likely that duplicate reads will be sampled. As a result, the true duplicate rate is not independent of the depth, and they should both be considered when looking at the duplicate rate. Additionally, as the sequencing depth in increases, it is also increasingly likely that reads will map to the same location and be marked as duplicates, even when they are not. As such, as the sequencing depth approaches and surpasses the read length, the duplicate rate starts to become less indicative of problems.
                              </div>
                              <div class="row vertical-align">
                                <div class="col-xs-4">
                                  <img title="Acceptable duplicate rate" style="width:100%; padding-bottom:15px;"  src="images/dup_good.png"></img>
                                </div>
                                <div class="col-xs-8">
                                  This is an example of the duplicate rate for a ~80X human whole genome. The expectation is that the duplicate rate is low (well below 10%), and consequently, this sample would be considered good.
                                </div>
                              </div>

                              <div class="row vertical-align">
                                <div class="col-xs-4">
                                  <img title="Acceptable duplicate rate" style="width:100%;padding-bottom:15px;"  src="images/dup_good_low_cov.png"></img>
                                </div>
                                <div class="col-xs-8">
                                  If the median coverage drops to ~50X, the duplicate rate should be even lower.
                                </div>
                              </div>

                              <div class="row vertical-align">
                                <div class="col-xs-4">
			          <img title="Potentially problematic duplicate rate" style="width:100%;"  src="images/dup_bad.png"></img>
                                </div>
                                <div class="col-xs-8">
                                  This is a different sample with ~50X coverage, but now the duplicate rate is much higher. This sample could well have problems at the library prep stage and should potentially be resequenced.
                                </div>
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>
                  </div>

                  <!-- Distributions -->
                  <div id="distributions" >

                      <!-- Read coverage distribution chart -->
                      <div id="read-coverage-distribution" class="distribution panel">
                        <div>Read Coverage Distribution
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="read_coverage_distribution_help" title="Expect a Poisson distribution centred on the expected mean coverage"></span>
                        </div>
                        <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                        <div id="read-coverage-distribution-chart" class="chart focus" preserveAspectRatio="none" style="width:98%;height:90%"></div>
                      </div>

                      <div id="length-distribution" class="distribution panel">
                        <div>
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="fragment_help" title="Expect a normal distribution"></span>
                          <span class="chart-chooser">
                            <span class="selected" onclick="toggleChart(this,'lengthChart')" data-outlier="false" data-id="frag_hist">Fragment Length</span> | <span onclick="toggleChart(this,'lengthChart')" data-id="length_hist" data-outlier="true">Read Length</span>
                          </span>
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="length_help" title="Expect an extremely narrow distribution"></span>
                        </div>
                        <label class="checkbox" for="checkbox2" style="position:absolute;right:10px;top:24px;cursor:pointer" >
                          <input type="checkbox"value="" class="outlier" data-toggle="checkbox" >
                            Outliers
                        </label>
                        <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                        <div id="length-distribution-chart" class="chart focus" preserveAspectRatio="none" style="width:98%;height:85%"></div>
                      </div>
                      <div id="mapping-quality-distribution" class="distribution panel">
                        <div>
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="map_qual_help" title="Expect distribution weighted to high values (~60)"></span>
                          <span class="chart-chooser">
                            <span onclick="toggleChart(this,'qualityChart')" data-id="mapq_hist" class="selected">Mapping Quality</span> | <span data-id="baseq_hist" onclick="toggleChart(this,'qualityChart')">Base Quality</span>
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="base_qual_help" title="Expect most values >40"></span>
                        </div>
                        <div class="samplingLoader">Sampling <img src="images/loading_dots.gif"/></div>
                        <div id="mapping-quality-distribution-chart" class="chart focus" preserveAspectRatio="none" style="width:98%;height:90%"></div>
                      </div>

                      <!-- Read coverage distribution modal -->
                      <div id="readCoverageDistributionModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Read coverage distribution</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                This chart shows how read coverage is distributed, and the expected distribution is dependent on the type of sequencing data being visualized.
                              </div>
                              <br>
                              <div><strong><i>Whole genome sequencing</i></strong></div>
                              <div>
                                In a whole genome sequencing experiment, the expectation is that the read coverage follows a Poisson distribution centred about the requested sequencing depth. The following example shows a high quality read coverage distribution for a sample sequenced to ~50X coverage. The distribution shows a nice Poisson distribution, and is centred around ~53X. (Note that the second scale at the bottom of the chart can be used to zoom in on desired parts of the distribution).
                                <img title="A good read coverage distribution" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/read_coverage_dist_zoomed.png"></img>
                                  Alternatively, if the distribution shows multiple peaks, isn't Poisson distributed, or is not centred around the expected coverage, it may be necessary to consider resequencing the sample, or at least, being aware that problems may arise in analysing the data. While the following distribution shows a median coverage around that expected (~80X), but with a significant portion of the genome at zero coverage and the multiple peaks, this would not be considered a good sample.
                                <img title="A bad read coverage distribution" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/read_coverage_dist_bad.png"></img>
                              </div>
                              <br>
                              <div><strong><i>Whole exome sequencing</i></strong></div>
                              <div>
                                Exome sequencing relies on the targetted capture of DNA from the exome, followed by DNA amplification. This leads to large variation in the sequencing depth across exons, and consequently, the read coverage distribution is no longer expected to be Poisson distributed. When sampling across the entire genome, the majority of genomic regions will contain no sequencing reads as will are not exonic regions. This leads to a read coverage distribution overwhelmingly weighted to zero coverage as shown below.
                                <img title="Exome sequencing" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/read_coverage_dist_exome.png"></img>
                                To restrict sampling to exonic regions, select the default bed file in the top 'Read Coverage' chart. It is also possible to select a custom bed file, if available. After selecting the default bed, the distribution above is updated to as shown below.
                                <img title="Exome sequencing using default bed" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/read_coverage_dist_exome_bed.png"></img>
                                This does not have a Poisson distribution and shows the wide distribution of coverage in the exonic regions. The sequenced depth appears to be centred around ~50X, so if this is consistent with the requested depth, this sample would be considered good.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Fragment length modal -->
                      <div id="fragmentModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Fragment length distribution</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                For paired end sequencing, DNA fragments are typically size selected to a uniform length and then sequenced from either end. Once the two mates are aligned back to the reference genome, the fragment length can be inferred from how far apart these two mates map. If the sequenced sample has a deletion or insertion relative to the reference, this will result in the two mates mapping closer together, or further apart than expected. Under the assumption that the sequenced sample has a relatively small number of insertions and deletions, we expect to see the fragment length follow a normal distribution.
                              </div>
                              <br>
                              <div><strong><i>Whole genome sequencing</i></strong></div>
                              <div>
                                This is an example of the fragment length distribution for a high coverage (~80X) whole genome. The read lengths in this sample are 150bp, so a fragment can not be shorter than this value, consequently, we see a sharp cutoff at a fragment length of 150bp.
                              </div>
                              <div>
                                <img title="WGS sequencing" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/fragment_wgs.png"></img>
                              </div>
                              <br>
                              <div><strong><i>Whole exome sequencing</i></strong></div>
                              <div>
                                Similarly, this is the fragment length distribution for a high coverage exome.
                              </div>
                              <div>
                                <img title="Exome sequencing" style="width:100%; padding-top:15px; padding-bottom:20px;"  src="images/fragment_exome.png"></img>
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Read length modal -->
                      <div id="lengthModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Read length distribution</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                The read length is usually a very simple distribution. In most cases, the read length is fixed at a uniform length, e.g. 100 base pairs, or 150 base pairs etc. The read length distribution, therefore, tends to be a single spike at this read length. Depending on the sequencing technology used, this may not always be the case.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Mapping quality distribution modal -->
                      <div id="mapQualModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Mapping quality distribution</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                The mapping quality distribution shows the Phred quality scores describing the probability that a read <i>does not</i> map to the location that it has been assigned to (specifically, Q=-log<sub>10</sub>(P), where Q is the Phred score and P is the probability the read is in the wrong location). So the larger the score, the higher the quality of the mapping. Some scores have specific meaning, e.g. a score of 0 means that the read could map equally to multiple places in the reference genome. The majority of reads should be well mapped and so we expect to see this distribution heavily skewed to large value (typically around 60). It is not unusual to see some scores around zero. Reads originating from repetitive elements in the genome will plausibly map to multiple locations.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>

                      <!-- Base quality distribution modal -->
                      <div id="baseQualModal" class="modal fade" role="dialog">
                        <div class="modal-dialog" style="width:700px;">
                          <div class="modal-content">
                            <div class="modal-header">
                              <button type="button" class="close" data-dismiss="modal">&times;</button>
                              <h4 class="modal-title">Base quality distribution</h4>
                            </div>
                            <div class="modal-body" style="overflow:auto;">
                              <div>
                                Similar to the mapping quality distribution, the base quality distribution shows the Phred quality scores describing the probability that a nucleotide has been <i>incorrectly</i> assigned; e.g. an error in the sequencing. Specifically, Q=-log<sub>10</sub>(P), where Q is the Phred score and P is the probability the nucleotide is wrong. The larger the score, the more confident we are in the base call. Depending on the sequencing technology, we can expect to see different distributions, but we expect to see a distribution skewed towards larger (more confident) scores; typically around 40.
                              </div>
                            </div>
                            <div class="modal-footer">
                              <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                            </div>
                          </div>
                        </div>
                      </div>
                  </div>
              </section>
          </div>
        </div>
        <script src="js/flatui-checkbox.js"></script>
        <script>
            var div = d3.select("body").append("div")
                .attr("class", "tooltip")
                .style("opacity", 1);
        </script>
    </body>
</html>
